linux/devdoc/async_socket_linux_requirements.md

```diff
@@ -275,62 +233,58 @@ MOCKABLE_FUNCTION(, int, async_socket_receive_async, ASYNC_SOCKET_HANDLE, async_
+    - **SRS_ASYNC_SOCKET_LINUX_11_089: [** If `errno` is `EAGAIN` or `EWOULDBLOCK`, then no data is available and `event_complete_callback` will break out of the function. **]**
 
-- **SRS_ASYNC_SOCKET_LINUX_11_087: [** Then `thread_worker_func` shall call `recv` and do the following: **]**
+    - **SRS_ASYNC_SOCKET_LINUX_11_090: [** If `errno` is `ECONNRESET`, then `event_complete_callback` shall call the `on_receive_complete` callback with the `on_receive_complete_context` and `ASYNC_SOCKET_RECEIVE_ABANDONED`. **]**
 
-- **SRS_ASYNC_SOCKET_LINUX_11_088: [** If the recv size < 0, then: **]**
+    - **SRS_ASYNC_SOCKET_LINUX_11_095: [** If `errno` is any other error, then `event_complete_callback` shall call the `on_receive_complete` callback with the `on_receive_complete_context` and `ASYNC_SOCKET_RECEIVE_ERROR`. **]**
 
-  - **SRS_ASYNC_SOCKET_LINUX_11_089: [** If `errno` is `EAGAIN` or `EWOULDBLOCK`, then unlikely errors will continue **]**
+  - **SRS_ASYNC_SOCKET_LINUX_11_091: [** If the recv size equal 0, then `event_complete_callback` shall call `on_receive_complete` callback with the `on_receive_complete_context` and `ASYNC_SOCKET_RECEIVE_OK`. **]**
 
-  - **SRS_ASYNC_SOCKET_LINUX_11_090: [** If `errno` is `ECONNRESET`, then `thread_worker_func` shall call the `on_receive_complete` callback with the `on_receive_complete_context` and `ASYNC_SOCKET_RECEIVE_ABANDONED` **]**
+  - **SRS_ASYNC_SOCKET_LINUX_11_092: [** If the recv size > 0, if we have another buffer to fill then we will attempt another read, otherwise we shall call `on_receive_complete` callback with the `on_receive_complete_context` and `ASYNC_SOCKET_RECEIVE_OK`. **]**
 
-- **SRS_ASYNC_SOCKET_LINUX_11_091: [** If the recv size equal 0, then `thread_worker_func` shall call `on_receive_complete` callback with the `on_receive_complete_context` and `ASYNC_SOCKET_RECEIVE_OK` **]**
+- **SRS_ASYNC_SOCKET_LINUX_11_093: [** `event_complete_callback` shall then free the `io_context` memory. **]**
 
-- **SRS_ASYNC_SOCKET_LINUX_11_092: [** If the recv size > 0, if we have another buffer to fill then we will attempt another read, otherwise we shall call `on_receive_complete` callback with the `on_receive_complete_context` and `ASYNC_SOCKET_RECEIVE_OK` **]**
+**SRS_ASYNC_SOCKET_LINUX_11_094: [** If the events value contains `COMPLETION_PORT_EPOLL_EPOLLOUT`, `event_complete_callback` shall the following: **]**
 
-- **SRS_ASYNC_SOCKET_LINUX_11_093: [** The `ASYNC_SOCKET_RECV_CONTEXT` object shall be removed from list of stored pointers. **]**
+- **SRS_ASYNC_SOCKET_LINUX_11_096: [** `event_complete_callback` shall call `send` on the data in the `ASYNC_SOCKET_SEND_CONTEXT` buffer. **]**
 
-**SRS_ASYNC_SOCKET_LINUX_11_094: [** If the events value contains `EPOLLOUT` (send), `thread_worker_func` shall the following: **]**
+- **SRS_ASYNC_SOCKET_LINUX_11_097: [** If send returns value is < 0 `event_complete_callback` shall do the following: **]**
```

Comment:
>send [](http://example.com/codeflow?start=42&length=4)

strictly talking about "send" it looks like 

`RETURN VALUE
       On success, these calls return the number of bytes sent.  On error, -1 is returned, and errno is set appropri‐
       ately.`

So what happens if we try to "send" 1MB of data, and send returns 100KB... what happens to the rest of the data??

... this is a reason to wait because we will absolutely have buffers big enough that we cannot send in 1 shot, but would otherwise be "send"-able in multiple "send"s.

Reply:
The code does this, but I didn't write the spec.  I will update.